service/qos/service_level_controller.cc

```diff
@@ -435,6 +441,80 @@ void service_level_controller::upgrade_to_v2(cql3::query_processor& qp, service:
     }
 }
 
+future<> service_level_controller::migrate_to_v2(size_t nodes_count, cql3::query_processor& qp, service::raft_group0_client& group0_client, abort_source& as) {
```

Comment:
If we already migrated data from legacy tables to the raft-based tables then we should skip this completely, right? Consider the following situation:

1. There is a cluster operating in legacy mode, having some non-trivial SL-related data,
2. The cluster is upgraded to SLs on raft and data is migrated,
3. Over time, some old SLs are dropped and some new ones are created (in SLs on raft mode),
4. A disaster happens and we need to perform recovery,
5. The last step of the recovery procedure is to redo the upgrade - considering that we don't clean up data in the old tables, it will overwrite the current raft state with the old SL data.

It would be great if we had a test for this - you can modify `test_service_levels_work_during_recovery` to start in legacy mode via error injection and adjust it accordingly.

Reply:
extended the test. I've added a service level to old table manually doing `INSERT INTO system_distributed.service_levels ...` and I'm checking if the service level is not migrated to v2 table after restore procedure end